### PR TITLE
Feature/improve search options

### PR DESCRIPTION
**Pull request by trasher**
See #1396

### CODE REVIEW

**Comment by trasher** [comment]
For reference, this is the patch I applied to get existing search options from master (commited in 39ef169):

```diff
diff --git a/tests/SearchTest.php b/tests/SearchTest.php
index a65c003..3442e55 100644
--- a/tests/SearchTest.php
+++ b/tests/SearchTest.php
@@ -229,7 +229,9 @@ class SearchTest extends DbTestCase {
             continue;
          }
          $item = getItemForItemtype($itemtype);
-         foreach ($item->getSearchOptions() as $key=>$data) {
+         $options = $item->getSearchOptions();
+         file_put_contents(__DIR__ . '/search_options/' . $itemtype, var_export($options, true));
+         foreach ($options as $key=>$data) {
             if (is_int($key)) {
                $input = array(
                    'itemtype' => $itemtype,
@@ -311,4 +313,4 @@ class SearchTest extends DbTestCase {
          $this->assertNotCount(0, $data['data'], $data['last_errors']);
       }
    }
-}
\ No newline at end of file
+}
```

File: inc/ticket.class.php [hunk 1]
```diff
@@ -2571,18 +2707,16 @@ function getSearchOptions() {
             $tokeep[] = 'validation';
          }
          $keep = false;
-         foreach ($tab as $key => $val) {
-            if (!is_array($val)) {
-               $keep = in_array($key, $tokeep);
+         foreach ($tab as $key => &$val) {
+            if (count($val) == 2) {
```
Comment (orthagh) [inline inc/ticket.class.php]: if think testing the presence (or absence) of field/table key could be preferable

File: inc/user.class.php [hunk 2]
```diff
@@ -2544,251 +2544,381 @@ static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBT
...
+         'forcegroupby'       => true,
+         'usehaving'          => true,
+         'datatype'           => 'count',
+         'massiveaction'      => false,
+         'joinparams'         => [
+            'beforejoin'         => [
+               'table'              => 'glpi_tickets_users',
+               'joinparams'         => [
+                  'jointype'           => 'child',
+                  'condition'          => 'AND NEWTABLE.`type`
```
Comment (orthagh) [inline inc/user.class.php]: CS

File: inc/user.class.php [hunk 3]
```diff
@@ -2544,251 +2544,381 @@ static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBT
...
```
Comment (orthagh) [inline inc/user.class.php]: CS

File: inc/authldap.class.php [hunk 4]
```diff
@@ -811,178 +811,268 @@ function defineTabs($options=array()) {
...
+         'name'               => __('Login field'),
+         'massiveaction'      => false,
+         'datatype'           => 'string'
+      ];
+
+      $tab[] = [
+         'id'                 => '9',
+         'table'              => $this->getTable(),
+         'field'              => 'realname_field',
+         'name'               => __('Last Name'),
```
Comment (trasher) [inline inc/authldap.class.php]: No, I surely missed that. That is strange test do no have fail on this one.

Comment (trasher) [inline inc/authldap.class.php]: (but maybe this is not part of tests...à

File: inc/cartridgeitem.class.php [hunk 5]
```diff
@@ -267,128 +267,187 @@ function getSpecificMassiveActions($checkitem=NULL) {
...
+         'field'              => 'id',
+         'name'               => __('Number of used cartridges'),
+         'datatype'           => 'count',
+         'forcegroupby'       => true,
+         'usehaving'          => true,
+         'massiveaction'      => false,
+         'joinparams'         => [
+            'jointype'           => 'child',
+            'condition'          => 'AND NEWTABLE.`date_use` IS NOT NULL
+                                                            AND NEWTABLE.`date_out` IS NULL'
```
Comment (trasher) [inline inc/cartridgeitem.class.php]: There is not real CS on that cases :p

File: inc/commonitilcost.class.php [hunk 6]
```diff
@@ -90,137 +90,192 @@ static function displayTabContentForItem(CommonGLPI $item, $tabnum=1, $withtempl
...
+         'field'              => 'id',
+         'name'               => __('ID'),
+         'massiveaction'      => false,
+         'datatype'           => 'number'
+      ];
+
+      $tab[] = [
+         'id'                 => '16',
+         'table'              => $this->getTable(),
+         'field'              => 'cost_material',
```
Comment (trasher) [inline inc/commonitilcost.class.php]: Will check.

File: inc/commonitilobject.class.php [hunk 7]
```diff
@@ -2717,142 +2717,211 @@ static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBT
...
-      $tab[1]['searchtype']           = 'contains';
-      $tab[1]['massiveaction']        = false;
-      $tab[1]['additionalfields']     = array('id', 'content', 'status');
-
-      $tab[21]['table']               = $this->getTable();
-      $tab[21]['field']               = 'content';
-      $tab[21]['name']                = __('Description');
-      $tab[21]['massiveaction']       = false;
-      $tab[21]['datatype']            = 'text';
+      $tab = [];
```
Comment (trasher) [inline inc/commonitilobject.class.php]: Yes, but I'm not sure we really want to keep those ones (will require much more work).

File: inc/ticket.class.php (hunk 1 above)
Comment (trasher) [inline inc/ticket.class.php]: Probably, I did not know what was really required for non labels entries :)

**Comment by trasher** [comment]
> what the purpose of test/search_options folder ?

To test old way vs new way output. For all files present in there, in the conditions of unit tests; we can be sure output has not been changed.

File: inc/commonitilobject.class.php (hunk 7 above)
Comment (orthagh) [inline inc/commonitilobject.class.php]: ok, thinking about it, it's not the "Breaking change".
We can pass.

File: inc/commonitilobject.class.php [hunk 8]
```diff
@@ -2717,142 +2717,211 @@ static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBT
...
+         'table'              => 'glpi_users',
+         'field'              => 'name',
+         'linkfield'          => 'users_id_lastupdater',
+         'name'               => __('Last edit by'),
+         'massiveaction'      => false,
+         'datatype'           => 'dropdown',
+         'right'              => 'all'
+      ];
+
+      // Filter search fields for helpdesk
```
Comment (trasher) [inline inc/commonitilobject.class.php]: If I remember well, this one was set elsewhere based on `$tab[45]`; I'll try to find where it was

Comment (trasher) [inline inc/commonitilobject.class.php]: was there: https://github.com/glpi-project/glpi/pull/1441/files/865f8f90436cac86a6d6b4b52921e311940578e8#diff-007edb12589cf3c8df3121cd7b19028aL2585

I do not see why this rule should only apply on tickets; that is why it has moved here. Was I wrong?

File: inc/commonitilcost.class.php (hunk 6 above)
Comment (trasher) [inline inc/commonitilcost.class.php]: It was a duplicate I did not see at the origin... So it's been removed :(

**Pull request merged**